File: packages/python/src/llmhub/router.py
from __future__ import annotations

from typing import List

from .types import ModelRecord, ModelResolutionRequest, ResolvedModel


class ModelRouter:
    def resolve(self, models: List[ModelRecord], request: ModelResolutionRequest) -> ResolvedModel:
        candidates = _filter_models(models, request)
        if not candidates:
            raise ValueError("router: no models match constraints")
        return ResolvedModel(
            primary=candidates[0],
            fallback=candidates[1:] if len(candidates) > 1 else None,
        )


def _filter_models(models: List[ModelRecord], request: ModelResolutionRequest) -> List[ModelRecord]:
    constraints = request.constraints
    preferred = [entry.strip() for entry in (request.preferredModels or []) if entry.strip()]
    allow_preview = constraints.allowPreview if constraints else True

    def matches(model: ModelRecord) -> bool:
        if not model.availability.entitled:
            return False
        if constraints:
            if constraints.requireTools and not model.features.tools:
                return False
            if constraints.requireJson and not (
                model.features.jsonMode or model.features.jsonSchema
            ):
                return False
            if constraints.requireVision and not model.modalities.vision:
                return False
            if constraints.maxCostUsd and not _within_cost(model, constraints.maxCostUsd):
                return False
        if allow_preview is False and model.tags and "preview" in model.tags:
            return False
        if preferred and not _matches_preferred(model, preferred):
            return False
        return True

    candidates = [model for model in models if matches(model)]
    candidates.sort(key=lambda m: (
        _preferred_rank(m, preferred),
        _price_score(m),
        m.displayName or "",
    ))
    return candidates


def _matches_preferred(model: ModelRecord, preferred: List[str]) -> bool:
    return any(entry in (model.id, model.providerModelId) for entry in preferred)


def _preferred_rank(model: ModelRecord, preferred: List[str]) -> int:
    if not preferred:
        return len(preferred) + 1
    for idx, entry in enumerate(preferred):
        if entry in (model.id, model.providerModelId):
            return idx
    return len(preferred) + 1


def _within_cost(model: ModelRecord, max_cost: float) -> bool:
    if not model.pricing:
        return True
    if model.pricing.inputPer1M and model.pricing.inputPer1M > max_cost:
        return False
    if model.pricing.outputPer1M and model.pricing.outputPer1M > max_cost:
        return False
    return True


def _price_score(model: ModelRecord) -> float:
    if not model.pricing:
        return 0.0
    inputs = model.pricing.inputPer1M or 0.0
    outputs = model.pricing.outputPer1M or 0.0
    if inputs and outputs:
        return min(inputs, outputs)
    return inputs or outputs

File: packages/node/src/core/router.ts
import { ModelRecord, ModelResolutionRequest, ResolvedModel } from "./types.js";

export class ModelRouter {
  resolve(models: ModelRecord[], request: ModelResolutionRequest): ResolvedModel {
    const candidates = filterModels(models, request);
    if (!candidates.length) {
      throw new Error("router: no models match constraints");
    }
    return {
      primary: candidates[0],
      fallback: candidates.length > 1 ? candidates.slice(1) : undefined,
    };
  }
}

function filterModels(
  models: ModelRecord[],
  request: ModelResolutionRequest,
): ModelRecord[] {
  const constraints = request.constraints ?? {};
  const preferred = (request.preferredModels ?? [])
    .map((entry) => entry.trim())
    .filter(Boolean);
  const allowPreview = constraints.allowPreview ?? true;
  const filtered = models.filter((model) => {
    if (!model.availability.entitled) {
      return false;
    }
    if (constraints.requireTools && !model.features.tools) {
      return false;
    }
    if (
      constraints.requireJson &&
      !(model.features.jsonMode || model.features.jsonSchema)
    ) {
      return false;
    }
    if (constraints.requireVision && !model.modalities.vision) {
      return false;
    }
    if (constraints.requireVideo && !model.modalities.videoOut) {
      return false;
    }
    if (!allowPreview && model.tags?.includes("preview")) {
      return false;
    }
    if (constraints.maxCostUsd && !withinCost(model, constraints.maxCostUsd)) {
      return false;
    }
    if (preferred.length && !matchesPreferred(model, preferred)) {
      return false;
    }
    return true;
  });
  return filtered.sort((a, b) => {
    const aRank = preferredRank(a, preferred);
    const bRank = preferredRank(b, preferred);
    if (aRank !== bRank) {
      return aRank - bRank;
    }
    const aCost = priceScore(a);
    const bCost = priceScore(b);
    if (aCost !== bCost) {
      return aCost - bCost;
    }
    return (a.displayName ?? "").localeCompare(b.displayName ?? "");
  });
}

function matchesPreferred(model: ModelRecord, preferred: string[]): boolean {
  return preferred.some(
    (entry) => entry === model.id || entry === model.providerModelId,
  );
}

function preferredRank(model: ModelRecord, preferred: string[]): number {
  if (!preferred.length) {
    return preferred.length + 1;
  }
  const index = preferred.findIndex(
    (entry) => entry === model.id || entry === model.providerModelId,
  );
  return index === -1 ? preferred.length + 1 : index;
}

function withinCost(model: ModelRecord, maxCost: number): boolean {
  const pricing = model.pricing;
  if (!pricing) {
    return true;
  }
  if (pricing.inputPer1M && pricing.inputPer1M > maxCost) {
    return false;
  }
  if (pricing.outputPer1M && pricing.outputPer1M > maxCost) {
    return false;
  }
  return true;
}

function priceScore(model: ModelRecord): number {
  if (!model.pricing) {
    return 0;
  }
  const input = model.pricing.inputPer1M ?? 0;
  const output = model.pricing.outputPer1M ?? 0;
  if (!input) {
    return output;
  }
  if (!output) {
    return input;
  }
  return Math.min(input, output);
}
